docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+1. `uniffi-bindgen-kotlin` asserts that it was compiled using the same `0.16` version, otherwise it panics.
+1. `uniffi-bindgen-kotlin` then calls a `generate_bindings` function in `uniffi-bindgen` (as a library dependency)
+1. `uniffi-bindgen`'s `generate_bindings` takes a generic parameter type that implements a trait `BindingGenerator`.
+1. `uniffi-bindgen` parses the `udl`, then calls the generic type's `write_bindings` function (the implementation lives in `uniffi-bingen-kotlin`)
+
+
+## **Implementation Option 2: Have `uniffi_bindgen` get dynamically linked with exactly one `uniffi_bindgen_{language}`**
+This approach uses the specific `uniffi_bindgen_{language}` as plugins. It essentially replaces the shelling out in `Implementation Option 1` with calling an exposed unsafe `C` API in a dynamically linked crate. 
+
+I'm not 100% sure how to let `uniffi_bindgen` choose which crate it gets dynamically linked with on run time, but it's possible.
+
+
+## **Implementation Option 3: Have each `uniffi_bindgen_{language}` be run directly**
+This approach has `uniffi_bindgen` only provide scaffolding and traits for each `uniffi_bindgen_{language}` to implement. `uniffi_bindgen` would also include some generic functions that can be called from `uniffi_bindgen_{language}` by treating `uniffi_bindgen` as a library dependency. The flow would look like this:
+
+1. A user would invoke `uniffi-bindgen-kotlin bindings -o <OUT_PATH> <UDL_PATH>`
+1. `uniffi-bindgen-kotlin` would call a generic function `generate_bindings` in `uniffi-bindgen` that takes type parameter `B: BindingGenerator`
+1. `uniffi-bindgen` parses the `udl`, then calls the generic type's `write_bindings` function (which lives in `uniffi-bingen-kotlin`)
+
+Version compatibility is a serious consideration with this approach, and we would benefit from exposing a `scaffolding` option to `uniffi-bindgen-kotlin`. This way the same `uniffi-bindgen-kotlin` is used to generate both scaffolding and bindings. We can also add a warning to `uniffi-bindgen`'s scaffolding that we recommend using a specific binary (like `uniffi-bindgen-kotlin`) to generate the scaffolding as well. Note: The scaffolding generating logic would still live in `uniffi-bindgen` we would only expose it to be called. `uniffi-bindgen-kotlin` would not modify the logic at all, this is purely for version compatibility.
+
+<br />
+<br />
+
+The old prototype in https://github.com/mozilla/uniffi-rs/pull/997 did Implementation Option 1, **I suggest we do Implementation Option 3 as its easier to reason about.**
+
+<br />
+<br />
+
+### Testing
+An important consideration for splitting the binding generators is the testing story for crates that live outside of the main `uniffi-rs` repository. We have a high-level decision to make:
+Should we expose our fixtures as testing standards? And if yes, how do we do that.
+I suggest that we leave this as an open question and open another issue to discuss it specifically.
```

Comment:
Yeah, I agree that this isn't worth blocking on if you've got a clear idea of how to move forward. My gut feeling is that publishing a deliberate test-fixtures crate is probably the right thing, but that's a whole lot of speculative work at this point.